Level 3

IrinaGarmaeva opened the pull request:
Realised functionality on front in accordance with the technical assignment

IrinaGarmaeva committed:
Initial commit

IrinaGarmaeva committed:
Feature: add Header and Footer components and styles

IrinaGarmaeva committed:
Feature: add Hero component

IrinaGarmaeva committed:
Feature: add About component

IrinaGarmaeva committed:
Feature: add Stack component

IrinaGarmaeva committed:
Feature: add Student component

IrinaGarmaeva committed:
Feature: add Portfolio component

IrinaGarmaeva committed:
Refactor Portfolio component

ove project to separate component

IrinaGarmaeva committed:
rename components due to tech. assignment

hero to promo, student to aboutme, about to about project, stack to techs

IrinaGarmaeva committed:
Make header responsive

IrinaGarmaeva committed:
Make Promo component responsive

IrinaGarmaeva committed:
Make responsive on 768px

IrinaGarmaeva committed:
Make responsive on mobile

IrinaGarmaeva committed:
Style: add styles on hover

IrinaGarmaeva committed:
Feature add Register component

IrinaGarmaeva committed:
Style: add styles to Register component

IrinaGarmaeva committed:
Feature add routing

IrinaGarmaeva committed:
Style: make Register component responsive

IrinaGarmaeva committed:
Feature: add Login component, make it responsive

IrinaGarmaeva committed:
Feature add SearchForm and FilterCheckBox components

make components responsive

IrinaGarmaeva committed:
Feature add MoviesCardList component

IrinaGarmaeva committed:
Style: make responsive MoviesCardList component

IrinaGarmaeva committed:
Feature: add NotFound component, make responsive

IrinaGarmaeva committed:
Feature: add Profile component, make responsive

IrinaGarmaeva committed:
Feature add navigation to separate component Navigation

IrinaGarmaeva committed:
Feature: add Button component, fix components display

display Header and Footer depending on route

IrinaGarmaeva committed:
Refactor move logo to separate component Logo

IrinaGarmaeva committed:
Refactor add Entry component for sign up/in form

IrinaGarmaeva committed:
Style add nav overlay on small screens

IrinaGarmaeva committed:
Ready to make pull request

IrinaGarmaeva committed:
Add alternative font

IrinaGarmaeva committed:
Edit Readme, fix style errors

IrinaGarmaeva committed:
Fix alt, semantics, links

IrinaGarmaeva committed:
Style: make nav overlay fixed

IrinaGarmaeva committed:
Fix sematics in SearchForm component

IrinaGarmaeva committed:
Feature add MainApi.js

create request to api, add functions to sign ip, sign in, chechToken

IrinaGarmaeva committed:
Feature: protect routes

add ProtectedRoute component to protect /movies, /saved-movies, /profile routes

IrinaGarmaeva committed:
Feature: add InfoPopup component

IrinaGarmaeva committed:
Feature add global state CurrentUserContext

IrinaGarmaeva committed:
Feature: add updateUserInfo

IrinaGarmaeva committed:
Feature add response messages to InfoPopup

depending on route and server response

IrinaGarmaeva committed:
Feature:add movies search by request/duration on route /movies

IrinaGarmaeva committed:
Feature: done Movies and SavedMovies components

add save and delete movie from favorites, display saved movies on /saved-movies route, save search data to local storage

IrinaGarmaeva committed:
Feature: add inputs validation, clean up code

vladko-89 reviewed (commented):
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно. **Комментарий: Поменять только почту возможности нет - кнопка заблокирована https://skr.sh/vLnqdedlJfj**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния. **Комментарий: Нажатие на кнопку не привязано к запросу. Т.е. даже если запрос будет завершен ошибкой, кнопка перекрасится.
    - **Комментарий: Должна быть возможность снять лайк на странице Фильмы**
  - На странице «Сохранённые фильмы»: 
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [ ] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт. **Комментарий: Если на странице Фильмы переключить чекбокс и вернуть обратно, далее перейти на другую страницу, а потом вернуться на Фильмы, в результате останется сообщение "Не найдено" https://skr.sh/vLnVIHIZgvw**
  - **Комментарий: Сбрасывать форму стоит только при успехе запроса, если запрос завершен с ошибкой, форму следует оставлять в открытом виде с внесенными данными**
  - **Комментарий: На странице Профиль из режима Редактирование нужно выходить после того как запрос завершился успехом. Сейчас пользователю придется снова менять данные, если пришла ошибка с сервера**
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [ ] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно. **Комментарий: У вас кнопка Назад возвращает на два шага в истории, хотя должна вернуть на предыдущий. Нужно понять в чем причина перерендера и устранить причину**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего. **Комментарий: Форму нужно блокировать на время запроса, чтобы у пользователя не было возможности отправлять повторные запросы до того как получен ответ сервераhttps://skr.sh/vLn7qDQuV9M Касается всех форм**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 53 (два балла сняты за Доп Комментарий)

## Доп. Комментарии:
- К фиксированным стоит отнести все числовые и строковые показатели, которые встречаются в коде "внезапно" - количество карточек которые нужно показать, количество добавляемых по кнопке Еще, ширины окна....
- Обращаю ваше внимание на макет. Если в макете есть блоки у которых фон упирается в макет, то в браузере нужно тянуть этот фон на всю ширину окна

IrinaGarmaeva committed:
Fix likes on /movies route

IrinaGarmaeva committed:
fix profile page

IrinaGarmaeva committed:
fixes after review

fix forms reset, back button on notfound page, disable inputs while request/search

vladko-89 reviewed (commented):
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.  
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния. **Комментарий: Нажатие на кнопку не привязано к запросу. Т.е. даже если запрос будет завершен ошибкой, кнопка перекрасится.**
  - На странице «Сохранённые фильмы»: 
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт. 
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [ ] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно. **Комментарий: У вас кнопка Назад возвращает на два шага в истории, хотя должна вернуть на предыдущий. Нужно понять в чем причина перерендера и устранить причину**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего. **Комментарий: Форму нужно блокировать на время запроса, чтобы у пользователя не было возможности отправлять повторные запросы до того как получен ответ сервераhttps://skr.sh/vLn7qDQuV9M Касается всех форм**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 60 (Один балл снят за Доп Комментарий)

## Доп. Комментарии:
 
- Обращаю ваше внимание на макет. Если в макете есть блоки у которых фон упирается в макет, то в браузере нужно тянуть этот фон на всю ширину окна

IrinaGarmaeva committed:
Fix like/dislike movie, back button on 404 page

vladko-89 reviewed (approved):
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.  
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.  
  - На странице «Сохранённые фильмы»: 
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт. 
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно. 
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [ ] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения. **Комментарий: Каждый переход на Сохраненные производит лишние запросы**
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего. **Комментарий: Кнопки нужно физически блокировать. Сейчас добавляется только css, что не мешает нажимать кнопку**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 95 (Один балл снят за Доп Комментарий)

## Доп. Комментарии:
 
- Обращаю ваше внимание на макет. Если в макете есть блоки у которых фон упирается в макет, то в браузере нужно тянуть этот фон на всю ширину окна

IrinaGarmaeva committed:
Edit Readme file
